openedx/core/lib/request_utils.py

```diff
@@ -103,162 +90,102 @@ def course_id_from_url(url):
-            names_to_size: Dict of sizes keyed by cookie name or cookie group name
-            top_n_captured: Number of largest sizes to monitor.
-            attribute_prefix: Prefix (cookies|cookies.group) to use in the custom attribute name.
-        """
-        top_n_cookies = sorted(
-            names_to_size,
-            key=lambda x: names_to_size[x],
-            reverse=True,
-        )[:top_n_captured]
-        for count, name in enumerate(top_n_cookies, start=1):
-            size = names_to_size[name]
-            name_attribute = f'{attribute_prefix}.{count}.name'
-            size_attribute = f'{attribute_prefix}.{count}.size'
-
-            set_custom_attribute(name_attribute, name)
-            set_custom_attribute(size_attribute, size)
-            log.debug('%s = %d', name, size)
+        is_large_cookie_header_detected = cookie_header_size >= logging_threshold
+        if not is_large_cookie_header_detected:
+            # .. setting_name: COOKIE_SAMPLING_REQUEST_COUNT
+            # .. setting_default: None
+            # .. setting_description: This setting enables sampling cookie header logging for cookie headers smaller
+            #   than COOKIE_HEADER_SIZE_LOGGING_THRESHOLD. The cookie header logging will happen randomly for each
+            #   request with a chance of 1 in COOKIE_SAMPLING_REQUEST_COUNT. For example, to see approximately one
+            #   sampled log message every 10 minutes, set COOKIE_SAMPLING_REQUEST_COUNT to the average number of
+            #   requests in 10 minutes.
+            # .. setting_warning: This setting requires COOKIE_HEADER_SIZE_LOGGING_THRESHOLD to be enabled to take
+            #   effect.
+            sampling_request_count = getattr(settings, "COOKIE_SAMPLING_REQUEST_COUNT", None)
+
+            # if the cookie header size is lower than the threshold, skip logging unless configured to do
+            #   random sampling and we choose the lucky number (in this case, 1).
+            if not sampling_request_count or random.randint(1, sampling_request_count) > 1:
+                return
+
+        # Sort starting with largest cookies
+        sorted_cookie_items = sorted(request.COOKIES.items(), key=lambda x: len(x[1]), reverse=True)
+        sizes = ', '.join(f"{name}: {len(value)}" for (name, value) in sorted_cookie_items)
+        if is_large_cookie_header_detected:
+            log_prefix = f"Large ({cookie_header_size} >= {logging_threshold}) cookie header detected."
+        else:
+            log_prefix = f"Sampled small ({cookie_header_size} < {logging_threshold}) cookie header."
+        log.info(f"{log_prefix} Cookie sizes: {sizes}")
+        # The computed header size can be used to double check that there aren't large cookies that are
+        #   duplicates in the original header (from different domains) that aren't being accounted for.
+        cookies_header_size_computed = sum([len(name) + len(value) + 2 for (name, value) in request.COOKIES.items()])
```

Comment:
You can drop the `[]` for sum if you'd like to, since it can take a generator as well. I also see 3 bytes per additional cookie (`; ` but also `=`) rather than two. You can then subtract two for the "missing fencepost" of one fewer inter-pair delimiter than cookie, if you want to get even closer. I think this should do it:

```suggestion
        cookies_header_size_computed = max(0, sum(len(name) + len(value) + 3 for (name, value) in request.COOKIES.items()) - 2)
```

I suspect the current unit tests are passing regardless because they always include two cookies.

Reply:
I was being sloppy. Thanks. Fixed and adjusted tests.